File: desktop/src/platform/updates.js
import pkg from "../../package.json";
import parseVersion from "semver/functions/parse.js";

const RELEASES_URL = "https://api.github.com/repos/sutao2/CueTuck/releases";

let testTransport = null;
let installTransport = null;

function isTauri() {
  return typeof window !== "undefined" && Boolean(window.__TAURI_INTERNALS__);
}

import { invokeCommand as tauriInvoke } from "./tauri.js";

export function normalizeUpdateChannel(channel) {
  return channel === "preview" ? "preview" : "stable";
}

export function resetUpdates() {
  testTransport = null;
  installTransport = null;
  Object.assign(updateState, { phase:'idle', version:'', notes:'', downloaded:0, total:null, error:'', checkedAt:'' });
}

export function setUpdateTransport(transport) {
  testTransport = transport;
}

export function setInstallTransport(transport) {
  installTransport = transport;
}

function fromReleases(releases, channel = "stable") {
  if (!Array.isArray(releases) || releases.length === 0) {
    return { available: false, notes: "" };
  }
  const wantPreview = channel === "preview";
  const candidates = releases.flatMap((row) => {
    const version = parseVersion(String(row?.tag_name ?? "").replace(/^v/i, ""));
    if (row?.draft || !version || Boolean(row?.prerelease) !== wantPreview || Boolean(version.prerelease.length) !== wantPreview) return [];
    return [{ row, version }];
  }).sort((a, b) => b.version.compare(a.version));
  const latest = candidates[0];
  if (!latest) {
    return { available: false, notes: "" };
  }
  const remote = String(latest.row.tag_name).replace(/^v/i, "");
  const current = parseVersion(String(pkg.version ?? ""));
  const available = Boolean(current) && latest.version.compare(current) > 0;
  return {
    available,
    notes: String(latest.row.body ?? ""),
    version: remote,
  };
}

export async function checkForUpdates({ channel } = {}) {
  const selected = normalizeUpdateChannel(channel);
  if (testTransport) return testTransport({ channel: selected });
  if (isTauri()) {
    return tauriInvoke("check_for_updates", { channel: selected });
  }
  const response = await fetch(`${RELEASES_URL}?per_page=100`, {
    headers: { Accept: "application/vnd.github+json" },
  });
  if (!response.ok) {
    throw new Error("检查失败");
  }
  return fromReleases(await response.json(), selected);
}

export async function queueUpdateInstall({ autoDownload, channel, version } = {}) {
  const selected = normalizeUpdateChannel(channel);
  if (!autoDownload) {
    return { queued: false, via: "updater" };
  }
  if (installTransport) {
    return installTransport({ channel: selected });
  }
  if (isTauri()) {
    return tauriInvoke("queue_update_install", { channel: selected, version });
  }
  return { queued: false, via: "updater" };
}

import { reactive } from 'vue';
import { getLocalSetting } from './library.js';
export const updateState = reactive({ phase: 'idle', version: '', notes: '', downloaded: 0, total: null, error: '', checkedAt: '' });
export const defaultUpdateChannel = pkg.version.includes('-') ? 'preview' : 'stable';
const busy = () => ['checking', 'downloading', 'verifying', 'installing'].includes(updateState.phase);
export async function selectedUpdateChannel() { return (await getLocalSetting('update_channel')) || defaultUpdateChannel; }
export async function refreshUpdate({ channel, autoDownload = false } = {}) {
  if (busy() || updateState.phase === 'ready') return;
  Object.assign(updateState, { phase: 'checking', error: '' });
  try {
    const result = await checkForUpdates({ channel: channel || await selectedUpdateChannel() });
    Object.assign(updateState, { phase: result.available ? 'available' : 'current', version: result.version || '', notes: result.notes || '', checkedAt: new Date().toLocaleString() });
    if (result.available && autoDownload) await downloadUpdate(channel);
  } catch (error) { Object.assign(updateState, { phase: 'error', error: error.message || '检查失败' }); }
}
export async function downloadUpdate(channel) {
  if (busy() || !updateState.version) return;
  if (!isTauri() && !installTransport) { updateState.error = '请在桌面客户端下载安装更新'; return; }
  Object.assign(updateState, { phase: 'downloading', downloaded: 0, total: null, error: '' });
  let unlisten;
  try {
    if (isTauri()) {
      const { listen } = await import('@tauri-apps/api/event');
      unlisten = await listen('update-progress', ({ payload }) => {
        if (['downloading','verifying'].includes(payload?.phase)) Object.assign(updateState, payload);
      });
    }
    const result = await queueUpdateInstall({ autoDownload: true, channel: channel || await selectedUpdateChannel(), version: updateState.version });
    if (!result?.ready) throw new Error('未能下载更新包，请重试');
    Object.assign(updateState, { phase: 'ready', downloaded: result.size, total: result.size });
  } catch (error) { Object.assign(updateState, { phase: 'available', error: error.message || '下载失败，请重试' }); }
  finally { unlisten?.(); }
}
export async function installUpdate() {
  if (updateState.phase !== 'ready') return;
  updateState.phase = 'installing'; updateState.error = '';
  try { await tauriInvoke('install_downloaded_update', { version: updateState.version }); }
  catch (error) { updateState.phase = 'ready'; updateState.error = String(error?.message || error); }
}
export function startUpdateChecks() {
  if (!isTauri()) return () => {};
  const check = async () => { try { if (!busy() && !['available','ready'].includes(updateState.phase)) await refreshUpdate({ autoDownload: (await getLocalSetting('auto_download')) === '1' }); } catch { /* A later check retries unavailable local preferences. */ } };
  const first = setTimeout(check, 10000);
  const timer = setInterval(check, 4 * 60 * 60 * 1000);
  return () => { clearTimeout(first); clearInterval(timer); };
}

File: desktop/src-tauri/src/commands/updates.rs
use reqwest::header::{HeaderMap, HeaderValue, USER_AGENT};
use serde_json::{json, Value};
use semver::Version;
use std::time::Duration;
use tauri::{AppHandle, Emitter, State};
use std::sync::Mutex;
use tauri_plugin_updater::Update;

#[derive(Default)]
pub struct UpdateState(pub Mutex<Option<(Update, Vec<u8>)>>, pub tokio::sync::Mutex<()>);
use tauri_plugin_updater::UpdaterExt;
use url::Url;

const RELEASES_URL: &str = "https://api.github.com/repos/sutao2/CueTuck/releases";

fn http_client() -> Result<reqwest::Client, String> {
    let mut headers = HeaderMap::new();
    headers.insert(
        USER_AGENT,
        HeaderValue::from_static("promptark-desktop"),
    );
    crate::http::apply_runtime_proxy(
        reqwest::Client::builder()
            .timeout(Duration::from_secs(10))
            .default_headers(headers),
    )?
    .build()
    .map_err(|_| "检查失败".to_string())
}

fn want_preview(channel: Option<&str>) -> bool {
    channel == Some("preview")
}

fn parse_tag(tag: &str) -> Option<Version> {
    Version::parse(tag.strip_prefix('v').or_else(|| tag.strip_prefix('V')).unwrap_or(tag)).ok()
}

fn pick_release(releases: &[Value], preview: bool) -> Option<&Value> {
    releases.iter().filter_map(|row| {
        let version = parse_tag(row.get("tag_name")?.as_str()?)?;
        if row.get("draft").and_then(Value::as_bool).unwrap_or(false)
            || row.get("prerelease").and_then(Value::as_bool).unwrap_or(false) != preview
            || !version.pre.is_empty() != preview { return None; }
        Some((row, version))
    }).max_by(|(_, a), (_, b)| a.cmp_precedence(b)).map(|(row, _)| row)
}

fn newer_than_current(latest: &Value) -> bool {
    let Some(remote) = latest.get("tag_name").and_then(Value::as_str).and_then(parse_tag) else { return false; };
    let Ok(current) = Version::parse(env!("CARGO_PKG_VERSION")) else { return false; };
    remote.cmp_precedence(&current).is_gt()
}

fn release_payload(latest: &Value) -> Value {
    let remote = latest
        .get("tag_name")
        .and_then(Value::as_str)
        .unwrap_or("")
        .trim_start_matches(['v', 'V']);
    let available = newer_than_current(latest);
    json!({
        "available": available,
        "notes": latest.get("body").and_then(Value::as_str).unwrap_or(""),
        "version": remote,
    })
}

async fn list_releases() -> Result<Vec<Value>, String> {
    let response = http_client()?
        .get(RELEASES_URL)
        .query(&[("per_page", "100")])
        .header("accept", "application/vnd.github+json")
        .send()
        .await
        .map_err(|_| "检查失败".to_string())?;
    if !response.status().is_success() {
        return Err("检查失败".to_string());
    }
    response.json().await.map_err(|_| "检查失败".to_string())
}

#[tauri::command]
pub async fn check_for_updates(channel: Option<String>) -> Result<Value, String> {
    let releases = list_releases().await?;
    let Some(latest) = pick_release(&releases, want_preview(channel.as_deref())) else {
        return Ok(json!({ "available": false, "notes": "" }));
    };
    Ok(release_payload(latest))
}

#[tauri::command]
pub async fn queue_update_install(app: AppHandle, channel: Option<String>, version: String, state: State<'_, UpdateState>) -> Result<Value, String> {
    let _guard = state.1.try_lock().map_err(|_| "更新正在进行".to_string())?;
    let releases = list_releases().await?;
    let latest = pick_release(&releases, want_preview(channel.as_deref())).ok_or("没有可用更新")?;
    if !newer_than_current(latest) || release_payload(latest)["version"] != version { return Err("版本已变化，请重新检查更新".into()); }
    let tag = latest["tag_name"].as_str().ok_or("版本无效")?;
    let endpoint = Url::parse(&format!("https://github.com/sutao2/CueTuck/releases/download/{tag}/latest.json")).map_err(|_| "更新地址无效")?;
    let mut builder = app.updater_builder().timeout(Duration::from_secs(180)).endpoints(vec![endpoint]).map_err(|_| "更新地址无效")?;
    if let Some(proxy) = crate::http::runtime_proxy_url()? { builder = builder.proxy(proxy); }
    let updater = builder.build().map_err(|_| "更新配置无效")?;
    let update = updater.check().await.map_err(|_| "读取更新包失败，请重试")?.ok_or("当前平台没有更新包")?;
    if update.version != version || update.download_url.scheme() != "https" || update.download_url.host_str() != Some("github.com") || !update.download_url.path().starts_with(&format!("/sutao2/CueTuck/releases/download/{tag}/")) {
        return Err("更新包与所选版本不一致".into());
    }
    *state.0.lock().map_err(|_| "更新状态不可用")? = None;
    let mut downloaded = 0u64;
    let mut last = std::time::Instant::now();
    let bytes = update.download(|chunk, total| {
        downloaded += chunk as u64;
        if last.elapsed() >= Duration::from_millis(100) || total == Some(downloaded) {
            let _ = app.emit("update-progress", json!({"phase":"downloading", "downloaded":downloaded,"total":total}));
            last = std::time::Instant::now();
        }
    }, || { let _ = app.emit("update-progress", json!({"phase":"verifying"})); }).await.map_err(|_| "下载或签名验证失败，请重试")?;
    let size = bytes.len();
    *state.0.lock().map_err(|_| "更新状态不可用")? = Some((update, bytes));
    Ok(json!({"ready":true,"version":version,"size":size}))
}

#[tauri::command]
pub async fn install_downloaded_update(app: AppHandle, version: String, state: State<'_, UpdateState>) -> Result<(), String> {
    let _guard = state.1.try_lock().map_err(|_| "更新正在进行".to_string())?;
    let staged = state.0.lock().map_err(|_| "更新状态不可用")?.clone().ok_or("请先下载更新")?;
    if staged.0.version != version { return Err("已下载版本不匹配，请重新下载".into()); }
    tauri::async_runtime::spawn_blocking(move || staged.0.install(staged.1)).await.map_err(|_| "安装任务失败")?.map_err(|_| "安装失败，请关闭其他实例后重试")?;
    app.restart();
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rejects_older_equivalent_and_invalid_versions_before_install() {
        for tag in ["v0.0.9".to_string(), format!("v{}+rebuild", env!("CARGO_PKG_VERSION")), "nightly".into(), "../latest".into()] {
            let release = json!({ "tag_name": tag });
            assert!(!newer_than_current(&release));
            assert_eq!(release_payload(&release)["available"], false);
        }
    }

    #[test]
    fn selects_highest_valid_non_draft_version_in_channel() {
        let releases = vec![
            json!({"tag_name": "v99.0.0", "draft": true}),
            json!({"tag_name": "nightly"}),
            json!({"tag_name": "v0.2.0"}),
            json!({"tag_name": "V0.10.0", "body": "newest"}),
            json!({"tag_name": "v1.0.0-beta.2", "prerelease": true}),
            json!({"tag_name": "v1.0.0-beta.10", "prerelease": true}),
            json!({"tag_name": "v2.0.0-beta.1", "prerelease": false}),
        ];
        let stable = release_payload(pick_release(&releases, false).unwrap());
        assert_eq!(stable, json!({"available": true, "version": "0.10.0", "notes": "newest"}));
        assert_eq!(pick_release(&releases, true).unwrap()["tag_name"], "v1.0.0-beta.10");
        assert!(pick_release(&[json!({"tag_name": "v2.0.0-beta.1"})], false).is_none());
        assert!(pick_release(&[], true).is_none());
    }
}
